src/output/timely_opt.rs

```diff
@@ -0,0 +1,330 @@
+                    let requires1 = borrow_region.clone();
+
+                    // requires(R2, B, P) :-
+                    //   requires(R1, B, P),
+                    //   subset(R1, R2, P).
+                    let requires2 = requires
+                        .map(|(r1, b, p)| ((r1, p), b))
+                        .join(&subset.map(|(r1, r2, p)| ((r1, p), r2)))
+                        .map(|((_r1, p), b, r2)| (r2, b, p));
+
+                    // requires(R, B, Q) :-
+                    //   requires(R, B, P),
+                    //   !killed(B, P),
+                    //   cfg_edge(P, Q),
+                    //   region_live_at(R, Q).
+                    let requires3 = requires
+                        .map(|(r, b, p)| ((b, p), r))
+                        .antijoin(&killed)
+                        .map(|((b, p), r)| (p, (r, b)))
+                        .join(&cfg_edge)
+                        .map(|(_p, (r, b), q)| ((r, q), b))
+                        .semijoin(&region_live_at)
+                        .map(|((r, q), b)| (r, b, q));
+
+                    requires1
+                        .concat(&requires2)
+                        .concat(&requires3)
+                        .distinct_total()
+                });
+
+                // .decl borrow_live_at(B, P) -- true if the restrictions of the borrow B
+                // need to be enforced at the point P
+                let borrow_live_at = {
+                    // borrow_live_at(B, P) :- requires(R, B, P), region_live_at(R, P)
+                    let borrow_live_at1 = requires
+                        .map(|(r, b, p)| ((r, p), b))
+                        .semijoin(&region_live_at)
+                        .map(|((_r, p), b)| (b, p));
+
+                    borrow_live_at1.distinct_total()
+                };
+
+                if dump_enabled {
+                    region_live_at.inspect_batch({
+                        let result = result.clone();
+                        move |_timestamp, facts| {
+                            let mut result = result.lock().unwrap();
+                            for ((region, location), _timestamp, multiplicity) in facts {
+                                assert_eq!(*multiplicity, 1);
+                                result
+                                    .region_live_at
+                                    .entry(*location)
+                                    .or_insert(vec![])
+                                    .push(*region);
+                            }
+                        }
+                    });
+
+                    subset.inspect_batch({
+                        let result = result.clone();
+                        move |_timestamp, facts| {
+                            let mut result = result.lock().unwrap();
+                            for ((r1, r2, location), _timestamp, multiplicity) in facts {
+                                assert_eq!(*multiplicity, 1);
```

Comment:
BTW this assert triggers with `-v``--stats` — apparently `multiplicity` being -1 here — with both input data sets.

Reply:
I just looked into this, and it seems that the problem goes away if you put a `consolidate()` after the definition of `subset`. This coalesces all updates together, and .. it seems to mean that the final results of `subset` are non-negative.

I think what happened here (testing now, but 5 min compile loop) is that 
```rust
                    let subset = Variable::from(subset_base.clone());
``` 
provides an initial value for `subset` and then removes it in the next round. The code concats the collection back in, so these two changes logically cancel, but the removal does produce negative tuples before they get cancelled.

I'm trying out 
```rust
                    let subset = Variable::from(subset_base.clone().filter(|_| false));
```
now which just says "don't start with any facts", which will avoid removals and shouldn't change the results as `subset_base` is concatenated in to the recursive definition of `subset` anyhow.

Reply:
Yes, adding in the filter to prevent initial values going in seems to prevent the assert, even after removing the `consolidate()` that I added. Now, technically speaking I haven't seen it assert so I'm trying that now to make sure that the code is actually being called (though using `--stats` and seeing stats outputs, so probably?).

Reply:
Yup, without the filter I do see the assert, and either the filter or consolidate seem to prevent it.

Reply:
Would you recommend I do this `filter` trick then? Or is `consolidate` better?

Reply:
(Could I just create an empty collection some other way...?)

Reply:
I'm sort of confused why we get the negative multiplicities, but I guess it's sort of an "internal accounting" thing?

Reply:
I think `filter` is probably "better". I forgot that with the `filter` you can drop the `clone`. It only really makes a difference if the input collection is large (the difference is one additional positive and negative instance of each input fact).

This is all here because `Variable` supports non-monotonic variables, and in principle a strictly monotonic variable shouldn't need any `from()` input.

Reply:
The negative multiplicities are really just about being general enough for non-monotonic computation. If you want to compute f^i(x) for a general f, we need to start with x, compute f(x), but then subtract x to get f(x) - x. That change is what needs to flow back around the loop to keep the incremental computation going correctly. 

In the case of Datalog you wouldn't need to subtract off the x, because you know you'll have distincts everywhere, but more general computation needs the -x to correctly determine f^i(x) rather than f(x + f(x + f(x + ...))).

Reply:
Also, yes I should totally have a better way to create an empty collection. Filter should be pretty cheap (the closure should optimize down to no per-record work). Probably the best other way is to use the same `to_stream` stuff you are using for inputs, but just with a `None` as the input `IntoIterator`.

Reply:
Yeah I figured I would do `None` or `vec![]` with an as-stream or whatever.